### PR TITLE
feat(geometry): add arc segment support to PolySlab with bulges

### PR DESCRIPTION
**Pull request by weiliangjin2021**
> [!NOTE]
> **Medium Risk**
> Touches core geometry computations (bounds, intersections, area/perimeter, containment) and introduces discretization-based behavior that could affect downstream meshing/queries, though guarded by strict validation and feature gating (no dilation/sidewall/adjoint).
> 
> **Overview**
> Adds arc-edge support to `PolySlab` via a new optional `bulges` array (bulge = tan(θ/4)) and propagates it through JSON schemas/changelog.
> 
> Implements arc geometry utilities, canonicalizes vertices+bulges together, and updates bounds/area/perimeter, `inside()`, and plane intersection logic to use an arc-discretized polygon when bulges are present, while enforcing new validations (length/finite values/no arc on zero-length edges/no self-intersecting arcs) and explicitly disallowing arcs with `dilation`, nonzero `sidewall_angle`, `interior_angle`, and adjoint derivatives. Extensive tests cover helper math, serialization, intersections, containment, and validation edge cases.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 40fccbfd828c2c3ab9c3276b5a6b8d85aecfab29. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: tidy3d/components/geometry/polyslab.py [hunk 1]
```diff
@@ -548,7 +1122,9 @@ def inside(self, x: NDArray[float], y: NDArray[float], z: NDArray[float]) -> NDA
 
             # vertical sidewall
             if math.isclose(self.sidewall_angle, 0):
-                face_polygon = shapely.Polygon(self.reference_polygon).buffer(fp_eps)
+                # Use discretized polygon for arc segments
+                poly_vertices = self._discretized_reference_polygon
```
Comment (yaugenst-flex) [inline tidy3d/components/geometry/polyslab.py:1122]: The ndarray path now uses `_discretized_reference_polygon`, but the scalar branch later in this method still uses `middle_polygon`/chord geometry (`vertices_z = self._shift_vertices(self.middle_polygon, dist)[0]`).

That makes `inside()` input-type dependent for bulged polygons: the same point can be inside for ndarray input and outside for scalar input (repro: `vertices=[(0,0),(1,0),(1,1),(0,1)]`, `bulges=[0.5,0,0,0]`, point `(0.5,-0.05,0.0)`). Could we make the scalar path arc-aware as well when `sidewall_angle == 0`?

**Review by yaugenst-flex — CHANGES_REQUESTED** [review]
I found one correctness issue: `inside()` gives inconsistent results for bulged polygons depending on input type (ndarray vs scalar). This conflicts with the intended arc semantics where discretization should be an implementation detail and results should match the arc-defined geometry.

Please align the scalar path with the arc-aware containment path and add a regression test that checks scalar/ndarray parity for the same point.

File: tidy3d/components/geometry/polyslab.py (hunk 1 above)
Comment (weiliangjin2021) [reply tidy3d/components/geometry/polyslab.py:1122]: Good catch, fixed now

**Comment by weiliangjin2021** [comment]
Some nice visualizations
bulge_varieties

File: tidy3d/components/geometry/polyslab.py [hunk 2]
```diff
@@ -104,6 +486,23 @@ class PolySlab(base.Planar):
         json_schema_extra={"units": MICROMETER},
     )
 
+    bulges: Optional[ArrayFloat1D] = Field(
```
Comment (lucas-flexcompute) [inline tidy3d/components/geometry/polyslab.py]: Ruff recommends me to use `bulges: ArrayFloat1D | None = ...` instead of `Optional`

File: tidy3d/components/geometry/polyslab.py [hunk 3]
```diff
@@ -72,13 +78,389 @@ def leggauss(n: int) -> tuple[NDArray, NDArray]:
     )
 
 
+class PolyBulgeUtil:
```
Comment (lucas-flexcompute) [inline tidy3d/components/geometry/polyslab.py]: If this is for internal use only, rename it to `_PolyBulgeUtil`, otherwise it's part of the public API.

File: tidy3d/components/geometry/polyslab.py [hunk 4]
```diff
@@ -72,13 +78,389 @@ def leggauss(n: int) -> tuple[NDArray, NDArray]:
...
+        arc_mask = bulge_data["arc_mask"]
+        if not np.any(arc_mask):
+            return float(np.sum(chord_lengths))
+
+        # For arc edges: compute arc lengths from precomputed data
+        radii = bulge_data["radii"]
+        arc_lengths = radii * np.abs(bulge_data["included_angles"])
+
+        # Replace chord lengths with arc lengths for arc edges
+        edge_lengths = chord_lengths.copy()
```
Comment (lucas-flexcompute) [inline tidy3d/components/geometry/polyslab.py]: Why copy if `chord_lengths` will not be reused? Maybe just call it `edge_lengths` from the start.

File: tidy3d/components/geometry/polyslab.py [hunk 5]
```diff
@@ -496,6 +988,88 @@ def _update_from_bounds(self, bounds: tuple[float, float], axis: Axis) -> PolySl
...
+        """Canonical (CCW, deduplicated) vertices and bulges as a synchronized pair."""
+        if self.bulges is None:
+            proper = self._proper_vertices(self.vertices)
+            return proper, np.zeros(proper.shape[0])
+        return self._canonicalize_vertices_and_bulges(self.vertices, self.bulges)
+
+    @cached_property
+    def _bulges(self) -> ArrayFloat1D:
+        """Postprocessed bulge values."""
+        return self._canonical_vertices_and_bulges[1]
```
Comment (lucas-flexcompute) [inline tidy3d/components/geometry/polyslab.py]: If `_canonical_vertices_and_bulges` is already cached, does this also need caching?

**Review by lucas-flexcompute — APPROVED** [review]
Is there a plan to implement sidewall angles and dilation with bulges? I'm guessing users will fall into that quite soon.

**Comment by weiliangjin2021** [comment]
> Is there a plan to implement sidewall angles and dilation with bulges? I'm guessing users will fall into that quite soon.

It'll be in the future.

File: tidy3d/components/geometry/polyslab.py (hunk 3 above)
Comment (weiliangjin2021) [reply tidy3d/components/geometry/polyslab.py]: good point, will rename it.
